### PR TITLE
Add support for fastmath 32-bit floating point divide

### PR DESCRIPTION
**Pull request by testhound**
This pull request adds support for generating more efficient code for 32-bit floating point divides in the presence of fastmath. This should be the final piece to: https://github.com/numba/numba/issues/6183.

### CODE REVIEW

File: numba/cuda/tests/cudapy/test_fastmath.py [hunk 1]
```diff
@@ -5,6 +5,7 @@
 
 
 class TestFastMathOption(CUDATestCase):
+    @skip_on_cudasim('fast divide not available in CUDASIM')
     def test_kernel(self):
```
Comment (gmarkall) [inline numba/cuda/tests/cudapy/test_fastmath.py:9]: For a while I was wondering what this is checking that differs from what `test_divf` checks for below... Eventually I concluded that it tests that casting an integer to a `float32` then using it in a division results in a `div.approx` instruction being emitted - given how finicky casts and the type system can be, I think it's good to keep this check alongside `test_divf` - Could you add a comment here explaining that this is testing the cast of an int being used in the divide, to make it a bit easier for future readers to see the difference?

Alternatively - do you have a different view on the difference between this and `test_divf`? I reached my conclusion by trying to derive meaning from the code (thinking backwards a bit) so maybe there's something else going on here that I haven't followed.

**Review by gmarkall — CHANGES_REQUESTED** [review]
I think the general approach here looks good - I have a few comments about the implementation and testing on the diff. I think with my suggested changes to the `maybe_fast_truediv` function, the following should be true:

- A division by zero involving `float32` operands in a kernel jitted with `debug=True` will raise a `ZeroDivisionError` following the launch.
- No exception should be raised following the launch of a kernel jitted with `fastmath=True, debug=True`.

If this is the case, could you add a test that tests both these conditions hold please?

**Review by gmarkall — APPROVED** [review]
Many thanks for the changes - they look good to me. Tests pass locally with hardware and the simulator.

I think the doc failure is unrelated to the contents of this PR - it is either a transient HTTP failure or an issue that needs resolving with the Numba docs.

@esc Could this have a buildfarm run please?

**Comment by esc** [comment]
Buildfarm ID: `numba_smoketest_cuda_yaml_20`

**Comment by gmarkall** [comment]
@esc How did it go? :-)

**Comment by stuartarchibald** [comment]
> @esc How did it go? :-)

Passed.

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch.

**Comment by stuartarchibald** [comment]
Not to resolve in this PR, but this raises an interesting question about flag compatibility, should fastmath and error model be tested to ensure compatibility?

**Pull request merged**